src/pages/resultsView/ResultsViewPageStore.ts

```diff
@@ -1304,61 +1304,66 @@ export class ResultsViewPageStore {
     );
 
     readonly molecularData = remoteData<NumericGeneMolecularData[]>({
-        await: () => [
-            this.studyToDataQueryFilter,
-            this.genes,
-            this.selectedMolecularProfiles,
-            this.samples,
-        ],
+        await: () => [this.samples, this.molecularData_preload],
         invoke: () => {
-            // we get mutations with mutations endpoint, structural variants and fusions with structural variant endpoint, generic assay with generic assay endpoint.
-            // filter out mutation genetic profile and structural variant profiles and generic assay profiles
-            const profilesWithoutMutationProfile = excludeSpecialMolecularProfiles(
-                this.selectedMolecularProfiles.result!
+            const samplesByUniqueKey = _.keyBy(
+                this.samples.result,
+                s => s.uniqueSampleKey
             );
-            const genes = this.genes.result;
+            return Promise.resolve(
+                this.molecularData_preload.result.filter(
+                    m => m.uniqueSampleKey in samplesByUniqueKey
+                )
+            );
+        },
+    });
 
-            if (
-                profilesWithoutMutationProfile.length &&
-                genes != undefined &&
-                genes.length
-            ) {
-                const identifiers: SampleMolecularIdentifier[] = [];
-
-                profilesWithoutMutationProfile.forEach(
-                    (profile: MolecularProfile) => {
-                        // for each profile, find samples which share studyId with profile and add identifier
-                        this.samples.result.forEach((sample: Sample) => {
-                            if (sample.studyId === profile.studyId) {
-                                identifiers.push({
-                                    molecularProfileId:
-                                        profile.molecularProfileId,
-                                    sampleId: sample.sampleId,
-                                });
-                            }
-                        });
-                    }
+    // loading molecular data based on profileId (instead of individual samples) provides
+    // two optimizations.
+    // 1. we can load this data before we know samples
+    // 2. backend can cache based on finite set of profiles
+    // we then have to filter this using samples, which can be loaded concurrently instead of serially
+    readonly molecularData_preload = remoteData<NumericGeneMolecularData[]>(
```

Comment:
How about `molecularDataForSelectedProfiles`? to make it same as how it is for gene panel data

Reply:
(I actually had to undo that optimization (molecularDataForSelectedProfiles) because in fact we need the data for ALL profiles and there was a test failing for coexpression.)

Do you mean molecularDataForAllProfiles_preload and molecularDataForAllProfiles?  I want to use the suffix because I don't believe this data should be used directly as it is not filtered by sample selection.

Reply:
ah okay, that makes sense. for the other one, that would be a one line fix to make the preload use all molecular profiles instead of selected molecular profiles